### ISSUE TITLE
Extend SVD gradient to support backpropagating through complex and (strongly) rectangular U and V

### ISSUE THREAD

**Issue by rmlarsen** [tensorflow/tensorflow#13641]
This initial version of SVD gradients has the following restrictions:
  Only supports statically known inner matrix dimensions m and n.

Backpropagating through U and V (i.e. backpropagating through SVD nodes with compute_uv=True) has further restrictions:
  a) Only supports real tensors.
  b) Only supports square and "almost square" matrices where the number of rows and columns differ by at most 1.
  c) full_matrices must be true also. This does not currently have severe implications, given the restriction in b).

Support for dynamic shapes and a) (I think) is straightforward to fix.  But b) is probably a deeper issue having to do with the (lack of) uniqueness of the decomposition, and will require some analysis. I think that if we understand b), we can get around the restriction in c) as well.

I'm marking this as contributions welcome, in the hope that somebody with better math skills than myself will help out :-)

**Comment by rmlarsen** [comment]
It looks like the implementation on autograd is better in several respects. That is probably a good starting point for working on this: https://github.com/HIPS/autograd/blob/master/autograd/numpy/linalg.py#L120

**Comment by guivenca** [comment]
This article studies the analytical computation of the Jacobian of the SVD decomposition, maybe it can be of use:
"Estimating the Jacobian of the Singular Value Decomposition : Theory and Applications " by Theodore Papadopoulo and Manolis L.A. Lourakis
https://hal.inria.fr/inria-00072686/document
Section 2.3.1 analyzes the degenerate case with repeated singular values. Their suggestion is to use a least square approach to obtain the "minimum norm" Jacobian.

That's probably why the finite difference gradient doesn't match the analytical one: even if it's able to compute a gradient, it may be only one of many.

**Comment by rmlarsen** [comment]
@guivenca thanks for the reference! I was thinking that such a minimal norm solution might be the right approach, but didn't work out the math. I'll study this paper.

**Comment by davharris** [comment]
@rmlarsen I just wanted to say that I really appreciate your work on this!

**Comment by vnavkal** [comment]
Would a "minimum norm" Jacobian actually be the one we need?  It seems to me that we need the Jacobian of the specific SVD function that's used.  I suspect this is an obstacle not only to removing the "almost square" restriction but also to extending support to complex tensors, since in the complex case there are always infinitely many choices of U and V.

Even in the real case, I could imagine the indeterminacy in U and V being a problem if someone tries to optimize over a function that depends on the signs of columns in U and V.  According to the paper @guivenca linked,
> ...even state of the art algorithms for SVD computation (eg LAPACK ’s dgesvd family of routines [1]) are “unstable” with respect to small perturbations of the input. By unstable, we mean that the signs associated with the columns and can change arbitrarily even with the slightest perturbation.

Anyway, it looks to me like it wouldn't be too hard to add support for `full_matrices=False`.  And I think it wouldn't need the restriction that the matrices are almost square.  If it seems useful, I'd be happy to work on a pull request for that; it would essentially be a port of the autograd code.

**Comment by rmlarsen** [comment]
@vnavkal I agree with your analysis. It would be super helpful if you would contribute a port of the autograd code. I would eventually get to it but am working on a few other projects right now.

**Comment by rmlarsen** [comment]
@davharris Thanks! I'm really hoping that we'll many more uses of sophisticated statistical model in TensorFlow with these types of algorithm available.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by JaeDukSeo** [comment]
any update on this matter?

**Comment by brianwa84** [comment]
I think rectangular SVD is already supported by the pull requests above. We have some work in progress on complex support, I'd guess it should be a few weeks to get it baked.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by zhuzhuy** [comment]
any update on complex?

**Comment by zhuzhuy** [comment]
I am confused that  I should write in both python backend and C++ backend  or just one backend is enough? (forgive my bad English)
Can anyone help me? I am really want to known

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by brianwa84** [comment]
No updates of interest thus far. We have the derivation for complex SVD reverse mode sensitivity, but have not yet implemented in TF, and have reduced the priority of that work.

**Comment by JaeDukSeo** [comment]
@brianwa84 just as a question, I have derived the back prop for eigenvalue decomposition is that even needed? Since svd is more stable I think it is not needed

**Comment by JaeDukSeo** [comment]
this is super complex work!

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by Randl** [comment]
> Would a "minimum norm" Jacobian actually be the one we need? It seems to me that we need the Jacobian of the specific SVD function that's used. 

As far as I understand, the idea is to through out the direction which is "parallel" to the family of the SVDs, since it will not move you to another family. While this is not a gradient, neither equals to numerical gradient, it could possibly work well. Also I think it will get you out of the region with undefined gradient.

> Even in the real case, I could imagine the indeterminacy in U and V being a problem if someone tries to optimize over a function that depends on the signs of columns in U and V.

> That the signs associated with the columns and can change arbitrarily even with the slightest perturbation.

I'm not sure why such an instability would exist. For distinct singular values we can always choose some convention for phase of `U`, for example, first non-zero element of each column is real positive, can't we?


Another idea I had is to try to determine the vectors in subspaces of degenerate singular values such that the we get terms of form 0/0 (i.e., if we have `F * M` (elementwise), and `F_ij = 1/0`, then `M_ij=0`), which then can be regarded as 0 to acquire meaningful gradient, similarly to [degenerated pertrubation theory](https://en.wikipedia.org/wiki/Perturbation_theory_(quantum_mechanics)#Effects_of_degeneracy). However, I didn't manage to acquire conditions on `U` columns in the subspace which would guarantee than (and not sure whether such conditions exist).

**Comment by MJfadeaway** [comment]
Any news?

**Comment by refraction-ray** [comment]
@brianwa84, sorry to bother. You mentioned that you guys had derived the backprop formula for complex valued SVD. Do you have any technical reports or notes? or just some hints or a direct formula on this? Since I am into this question recently and find that the formula should be very different from the real one due to the possible imaginary diagonals for $U^\dagge dU$. 

I didn't find many relevant useful resources on this topic. The derivation in [this](https://giggleliu.github.io/2019/04/02/einsumbp.html) definitely overlooked the diagonal part of $U^\dagger dU$, a very common loophole. In [this paper](https://arxiv.org/pdf/1907.13422.pdf), the author also left backprop for complex SVD unsolved.

Any help on this is appreciated. (I am also curious why tf team reduced the priority of the implementation when you have the derivations, I'd guess it is a super complex formula?)

**Comment by GiggleLiu** [comment]
> @brianwa84, sorry to bother. You mentioned that you guys had derived the backprop formula for complex valued SVD. Do you have any technical reports or notes? or just some hints or a direct formula on this? Since I am into this question recently and find that the formula should be very different from the real one due to the possible imaginary diagonals for $U^\dagger dU$.
> 
> I didn't find many relevant useful resources on this topic. The derivation in [this](https://giggleliu.github.io/2019/04/02/einsumbp.html) definitely overlooked the diagonal part of $U^\dagger dU$, a very common loophole. In [this paper](https://arxiv.org/pdf/1907.13422.pdf), the author also left backprop for complex SVD unsolved.
> 
> Any help on this is appreciated. (I am also curious why tf team reduced the priority of the implementation when you have the derivations, I'd guess it is a super complex formula?)

Interesting point, could you please show which part ignores the diagonal imaginary part in [this](https://giggleliu.github.io/2019/04/02/einsumbp.html) blog? This formula was tested in this repo: https://github.com/GiggleLiu/BackwardsLinalg.jl/blob/master/test/svd.jl . And tested in practice in this [GSoC project](https://github.com/under-Peter/TensorNetworkAD.jl).

You rightly that the loss function in these tests is carefully designed. it does not pass tests with a general U adjoint, as it shouldn’t due to the gauge freedom issue. If you can offer a **gauge independent** loss that breaks this test, it will be very helpful for me to fix the derivation.

**Comment by refraction-ray** [comment]
Hi @GiggleLiu, 
<img width="342"  src="https://user-images.githubusercontent.com/35157286/64085121-51d6d180-cd63-11e9-9e7b-2dcbdc60cfd4.png">
Note how the above formula is valid only for off-diagonal elements. There is one part of differentiation missing, namely the diagonal part $dC_{ii}$ (which is not zero in general but pure imaginary). Unless one can argue that under some gauge, $diag(dC)$ and $diag(dD)$ could be zero simultaneously or somehow cancelled with each other in dL formula.

Actually, we can show that if the loss function is a real gauge invariant function depending on U OR V only, then the formula above is enough (somehow the diagonal part of dC and dD can be cancelled in this case). However, we didn't have a general proof on why we don't need to consider the differentiation in diagonal part for dC and dD in the general case. If the loss function is some thing like `tf.real(tf.conj(V[0,0])*U[0,0])`,  then the numerical test seems failed. If I am understanding right, such an object function is gauge invariant (I'd guess the gauge freedom is some diagonal phase matrix P transforming U and V as UP and VP, or is there a larger gauge freedom?).

In sum, either the formula in your post is incomplete with one part missing, or we should have a general proof on why ignoring the diagonal part of dC and dD is safe.

Best

**Comment by GiggleLiu** [comment]
Hi, @refraction-ray 

I think you are right. Will look into it ASAP. Thanks!

PS: I am in IOP, CAS  (中科院物理所), welcome for visiting.

**Comment by GiggleLiu** [comment]
Some updates:

1. The test passes for `tf.real(tf.conj(V[0,0])*U[0,0])`, 
2. This statement in the blog is not correct

> where dCS and SdD has zero diagonal elements. So that dS=diag(dP).`dS=diag(dP)`

Should be `dS = real(diag(dP))`, the code is correct though. The rest are correct. Now the blog has been updated.

Nice observation and thanks for correction, @refraction-ray

**Comment by refraction-ray** [comment]
> Some updates:
> 
> 1. The test passes for `tf.real(tf.conj(V[0,0])*U[0,0])`,
> 2. This statement in the blog is not correct
> 
> > where dCS and SdD has zero diagonal elements. So that dS=diag(dP).`dS=diag(dP)`
> 
> Should be `dS = real(diag(dP))`, the code is correct though. The rest are correct. Now the [blog](https://giggleliu.github.io/2019/04/02/einsumbp.html) has been updated.
> Here, only non-zero real part is important to the final result, since `dS` must be real.
> 
> Nice observation and thanks for correction, @refraction-ray

Yes, for dS, we can still have the same derivation, but I am not sure about dC and dD part. What is the value for $$dD_{i,i}$$? The formula in the above figure captures only relation on off-diagonal parts (dC has no diagonal contribution to the equation). In other words, dC should be some thing as $dC=dC_0+dC_d$ where dC_0 in the same part at real SVD case, dC_d part is differentiations on diagonal imaginary parts (unless one can find a proof showing this is zero).

And I am also not sure about the numerical test, since I have some code snippets works for real case or complex case with object function $U[0,0]U*[0,0]$, but fails when loss function is $UV*$. (I will have a double check on this implementation part, though)

**Comment by GiggleLiu** [comment]
I see, indeed, the imaginary part of dC and dD is ignored. Is your test same as this one:

https://github.com/GiggleLiu/BackwardsLinalg.jl/blob/adfcad724273f0403e20ac4712fb7273cbac6f8d/test/svd.jl#L42

I our case, it does not error, you'd better check your program again.

I will look into the imaginary part later, it looks not trivil.

**Comment by refraction-ray** [comment]
Hi, @GiggleLiu. 

We have derived the correct backprop formula for squared SVD today with diagonal part of dC and dD. Their contributions are indeed nonzero. And we also tested the new formula numerically, which matches finite differentiation for `tf.real(tf.conj(V[0,0])*U[0,0])`. We will write up the whole thing soon, and will definitely email you the preprint for discussion before we publish the result.

Best

**Comment by GiggleLiu** [comment]
Congratuations! look forward to your derivation.

It is still not clear to me why the autodiff check passes in our Julia program for the same loss with incorrect formula. 😂  Hope after reading your post, I will be clear.

**Comment by refraction-ray** [comment]
Please check the note at https://re-ra.xyz/misc/complexsvd.pdf for the correct backprop formula of complex valued SVD. (I'd like to post this on arXiv for a better reference in the future, but somehow the submission entered onhold status :( and I decided to share this technical report with the community first.)

Also remember to check the updated [blog](https://giggleliu.github.io/2019/04/02/einsumbp.html) by @GiggleLiu , which now has the correct backprop formula and derivation for complex SVD.

**Comment by GiggleLiu** [comment]
A general remark, @refraction-ray just solves the notorious [gauge problem](https://arxiv.org/abs/1907.13422) in the bacpropgation of complex values SVD, making [some important applications](https://arxiv.org/abs/1903.09650) in physics can be reliablly use in general cases.

#### For those who are interested in gradient check
The reason why my gradient check fails: I checked `f(x-a*g) = f(x)-a*|g|^2`, which is not enough, it only gaurantes the training moves to the correct direction, but can loss some components, like a projection of a high dimensional vector to a low dimension `[1, 1] -> [1, 0]`. The missing part is what @refraction-ray finds. With this missing part, we can have a better differential programming quantum world.

**Comment by Randl** [comment]
So are we going to see it implemented in TF soon? :)

**Comment by GiggleLiu** [comment]
@Randl I think so.
@refraction-ray do you want to take this honor? Ping me if you need any help in submitting a pull request. 😄

**Comment by brianwa84** [comment]
Happy to help with reviewing/getting it pulled in.

**Comment by helo-nerd** [comment]
> @Randl I think so.
> @refraction-ray do you want to take this honor? Ping me if you need any help in submitting a pull request. 😄

Hi, can I ask If your new update is implemented in TF now already? It is very significant for my research.

**Comment by refraction-ray** [comment]
@helo-nerd , yes, I believe AD on complex valued SVD is ready for tensorflow 2.1+

**Comment by helo-nerd** [comment]
> @helo-nerd , yes, I believe AD on complex valued SVD is ready for tensorflow 2.1+

Ohh that's awesome! is this new update the same code list here? Because my server has some problem for TF 2, I wanna continuely use TF 1 by modifying it.
https://github.com/tensorflow/tensorflow/pull/32226/commits/124386cd8312c530097cf930be9cdb3544c1e361

**Comment by refraction-ray** [comment]
@helo-nerd , yes, the code should be the same as this commit, you can directly hack tf source as this commit, or use [``customize_gradient``](https://www.tensorflow.org/api_docs/python/tf/custom_gradient) in your app code.

**Comment by Randl** [comment]
Is it actually solved?

**Comment by brianwa84** [comment]
yes, was addressed in https://github.com/tensorflow/tensorflow/pull/32226
